utils/ami-cleanup/ami_cleanup.py

```diff
@@ -0,0 +1,220 @@
+                )
+                return True
+    except Exception as e:
+        logger.error(f"Failed to get list of Launch Configurations using AMI ID {ami_id} : {e}")
+        raise
+
+    # Check against launch templates
+    try:
+        lt_list = ec2_client.describe_launch_templates()["LaunchTemplates"]
+        for lt in lt_list:
+            lt_desc = ec2_client.describe_launch_template_versions(
+                LaunchTemplateId=lt["LaunchTemplateId"],
+                Versions=["$Latest"],
+                Filters=[{"Name": "image-id", "Values": [ami_id]}],
+            )["LaunchTemplateVersions"]
+            if len(lt_desc) > 0:
+                logger.debug(
+                    f"AMI {ami_id} is used by Launch Template {lt['LaunchTemplateName']}"
+                )
+                return True
+    except Exception as e:
+        logger.error(f"Failed to get list of Launch Templates using AMI ID {ami_id} : {e}")
+        raise
+
+    return False
+
+
+def delete_ami(ec2_client, ami_id, dry_run):
+    # Deregister AMI
+
+    try:
+        if not dry_run:
+            ec2_client.deregister_image(ImageId=ami_id)
+            logger.info(f"AMI {ami_id} deregistered")
+        else:
+            logger.info(f"AMI {ami_id} would be deregistered if this wasn't a dry run")
+    except Exception as e:
+        logger.debug(f"Error while deregistering AMI {ami_id}")
+        logger.debug(e)
+
+    # Get a list of snapshots associated with the ami
+    snap_filters = [template.replace("AMI_ID", ami_id) for template in snap_filter_templates]
+    snapshot_list = []
+
+    for filter in snap_filters:
+        try:
+            snapshot_list_for_filter = ec2_client.describe_snapshots(
+                Filters=[{"Name": "description", "Values": [filter]}]
+            )["Snapshots"]
+        except Exception as e:
+            logger.error(f"Failed to get list of Snapshots for AMI ID {ami_id} : {e}")
+            raise
+        if len(snapshot_list_for_filter) > 0:
+            snapshot_list.append(snapshot_list_for_filter)
+
+    if len(snapshot_list) > 0:
+        # Flatten list of lists generated above
+        flat_snapshot_list = []
+        for sublist in snapshot_list:
+            for item in sublist:
+                flat_snapshot_list.append(item)
+
+        for snapshot in flat_snapshot_list:
+            snapshot_id = snapshot["SnapshotId"]
+            try:
+                if not dry_run:
+                    ec2_client.delete_snapshot(SnapshotId=snapshot_id)
+                    logger.info(f"Snapshot {snapshot_id} deleted")
+                else:
+                    logger.info(
+                        f"Snapshot {snapshot_id} would be deleted if this wasn't a dry run"
+                    )
+            except Exception as e:
+                logger.debug(e)
+    else:
+        logger.error(f"No snapshots found for AMI {ami_id}")
```

Comment:
what should we do if there are no snapshots found?

Reply:
A situation where no snapshots exist for an AMI may arise where outside of this script, e.g. manually, a snapshot is deleted  but an AMI isn't deregistered. In such cases it is appropriate to log an error and continue.

Reply:
Might be worth adding that messages as a comment then